config.go

```diff
@@ -22,51 +25,112 @@ var (
 
 	// Commandline Version String; used to control auto-updating.
 	ClientVersion = "2017-01-18"
+
+	errNotFound = "not found"
 )
 
 // AuthUser configures a user for our Naive authentication setup.
 type AuthUser struct {
-	Username    string `yaml:"username"`
-	DisplayName string `yaml:"display_name"`
-	Password    string `yaml:"password"`
-	Email       string `yaml:"email"`
+	Username    string `bson:"username" json:"username" yaml:"username"`
+	DisplayName string `bson:"display_name" json:"display_name" yaml:"display_name"`
+	Password    string `bson:"password" json:"password" yaml:"password"`
+	Email       string `bson:"email" json:"email" yaml:"email"`
 }
 
 // NaiveAuthConfig contains a list of AuthUsers from the settings file.
 type NaiveAuthConfig struct {
-	Users []*AuthUser
+	Users []*AuthUser `bson:"users" json:"users" yaml:"users"`
 }
 
 // CrowdConfig holds settings for interacting with Atlassian Crowd.
 type CrowdConfig struct {
-	Username string
-	Password string
-	Urlroot  string
+	Username string `bson:"username" json:"username" yaml:"username"`
+	Password string `bson:"password" json:"password" yaml:"password"`
+	Urlroot  string `bson:"url_root" json:"url_root" yaml:"urlroot"`
 }
 
 // GithubAuthConfig holds settings for interacting with Github Authentication including the
 // ClientID, ClientSecret and CallbackUri which are given when registering the application
 // Furthermore,
 type GithubAuthConfig struct {
-	ClientId     string   `yaml:"client_id"`
-	ClientSecret string   `yaml:"client_secret"`
-	Users        []string `yaml:"users"`
-	Organization string   `yaml:"organization"`
+	ClientId     string   `bson:"client_id" bson:"client_id" yaml:"client_id"`
+	ClientSecret string   `bson:"client_secret" json:"client_secret" yaml:"client_secret"`
+	Users        []string `bson:"users" json:"users" yaml:"users"`
+	Organization string   `bson:"organization" json:"organization" yaml:"organization"`
 }
 
 // AuthConfig has a pointer to either a CrowConfig or a NaiveAuthConfig.
 type AuthConfig struct {
-	Crowd  *CrowdConfig      `yaml:"crowd"`
-	Naive  *NaiveAuthConfig  `yaml:"naive"`
-	Github *GithubAuthConfig `yaml:"github"`
+	Crowd  *CrowdConfig      `bson:"crowd" json:"crowd" yaml:"crowd"`
+	Naive  *NaiveAuthConfig  `bson:"naive" json:"naive" yaml:"naive"`
+	Github *GithubAuthConfig `bson:"github" json:"github" yaml:"github"`
+}
+
+func (c *AuthConfig) id() string { return "auth" }
+func (c *AuthConfig) get() error {
+	err := legacyDB.FindOneQ(Collection, legacyDB.Query(byId(c.id())), c)
+	if err != nil && err.Error() == errNotFound {
+		return nil
+	}
+	return err
+}
+func (c *AuthConfig) set() error {
+	_, err := legacyDB.Upsert(Collection, byId(c.id()), bson.M{
+		"$set": bson.M{
+			"crowd":  c.Crowd,
+			"naive":  c.Naive,
+			"github": c.Github,
+		},
+	})
+	return err
+}
+func (c *AuthConfig) validateAndDefault() error {
```

Comment:
This function can error in several ways. It might make sense to use a catcher, to avoid the case where someone fixes one problem only to discover another (duplicate user, no github auth).